chapel-lang/chapel
# Allow importing of nested modules using their full path

Allows the import of nested modules using the full path to the module, or a path starting at a module that has been brought into scope by a use or other import statement.  Does not allow import of a nested module relative to the current scope - e.g. for a module structure:
```chapel
module M {
  module N { ... }
  import M.N; // will work
  import N; // will not work
}
```

Resolves Cray/chapel-private#784

Details:
Alters the parser to allow expressions in import statements, rather than just identifiers.

As a consequence, adds a method to ImportStmts to validate that the expressions provided are allowed, and calls it when performing scope resolution.  I choose to make this similar to the one for UseStmts but separate for now, due to anticipated differences in the symbols that will be allowed in ImportStmts.

Alters how ImportStmts determine the module being imported, due to the old strategy not being able to easily distinguish between good accesses of nested modules and bad accesses.  Now ImportStmts use a separate strategy from use statements to determine their referenced modules - it starts from the current scope, only allowing local modules if they have been given a full path.  It follows use and import statements and allows nested modules found due to those statements without a full path.  Finally, it considers top level modules.

Tests:
Adds the following new tests:
- Checks importing a parent module that is also nested using the full path (In A.B.C, importing A.B, expected to work)
- Checks importing a parent module that is also nested without using the full path (In A.B.C, importing B, expected to fail)
- Importing a three deep module from outside its hierarchy using the full path (expected to pass)
- Importing a three deep module from its grandparent module, using a shorter path (to ensure that we don’t just allow it because there’s a `.` as part of the path)
- Importing a nested module, then using the imported module to import a module within it (expected to work)
- Using a module and then importing a module defined within the used module (expected to work)
- Importing a nested module and renaming it, using the new name to access its symbols
- Importing a nested module and renaming it, and using the new name to import a module it defines (despite the existence of a top level module with the same name as the new module, matching Rust’s behavior)
- Importing a nested module from outside its hierarchy and trying to access a function in an unqualified manner (expected to fail)
- Importing a nested module from outside its hierarchy and trying to access a module scoped variable in an unqualified manner (expected to fail)
- Importing a parent module and trying to access a function in an unqualified manner (works due to the function still being in scope, similar to if there were no uses or imports at all)
- Importing a top level module with the same name as a closer nested module (works, but generates a warning message)

And alters the following old tests:
- Updates the original test of full path nested module importing to demonstrate the intended proper strategy to access the symbols
- Update the error message on the first future for nested modules to use the more specific error message language
- Alter the expected error message for importing a literal to reflect it no longer being a syntax error (but still being an error)
- Alter the expected behavior for renaming a module to the same name as a top level module and then trying to import with that name
    - To allow using renamed modules to import other modules, and to follow Rust’s behavior of looking at the import first, this now matches to the renamed module twice.  While in Rust, this results in an error message about importing it twice, I choose to follow the behavior of UseStmts and not warn
- Update the .bad for using an import to access a module’s symbols to reflect `.` no longer being a syntax error (but still an error for now)

Testing:
- Full std paratest with futures (passed)
